### PR TITLE
sg: adds a command to check a release build for approved cves

### PR DESCRIPTION
**Pull request by coury-clark**
Automates a step in the release process (at least provides some utility) to compare CVEs found in a build to a set of CVEs that have been approved by the security team.

> Cross check all reported CVEs are in the accepted list (https://handbook.sourcegraph.com/departments/security/tooling/trivy/4-2-0). Otherwise, alert @security-support in the [#release-guild](https://sourcegraph.slack.com/archives/C032Z79NZQC) channel ASAP.

The implementation is simple
1. Pull all of the HTML security artifacts generated in the release and extract any CVE strings using a regexp 
2. Download a provided URL (typically a handbook page) 
3. Check the downloaded text for any usage of the CVE strings. It's not perfect, but it should get the job done for release captains

Usage is simple:

`go run ./dev/sg release cve-check -u https://handbook.sourcegraph.com/departments/security/tooling/trivy/4-2-0 -b 184191`

```
➜  sourcegraph git:(main) ✗ go run ./dev/sg release cve-check -u https://handbook.sourcegraph.com/departments/security/tooling/trivy/4-2-0 -b 184191
Checking release for approved CVEs...
Downloading url: https://handbook.sourcegraph.com/departments/security/tooling/trivy/4-2-0
Checking security report: indexed-searcher-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e450-45be-b119-3de3a7820657/artifacts/018486f4-ddbe-4ca1-8291-369628f05522/download
Checking security report: redis_exporter-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e45e-4672-9c34-2f1b3774abb0/artifacts/018486f4-fc3e-4649-8816-18d1b0222179/download
Checking security report: node-exporter-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e456-4c67-921c-3f5e46541d05/artifacts/018486f5-01a0-4acd-b3f3-9658da6bca98/download
Checking security report: redis-cache-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e45c-4709-9c36-9db42ef597da/artifacts/018486f5-040e-4b4a-a878-539e11d6fbb9/download
Checking security report: search-indexer-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e461-4ed2-a1ca-df5fc4d1b4da/artifacts/018486f5-0868-4cdd-a02a-7dd1d2182034/download
Checking security report: cadvisor-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e448-4a19-8177-0af61753b4c9/artifacts/018486f5-0cc2-4d65-a10a-cf804938fa59/download
Checking security report: jaeger-all-in-one-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e453-4ffe-b077-1e93e9aa5310/artifacts/018486f5-23c5-4af9-b8c1-638eadc65f67/download
Checking security report: batcheshelper-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e46a-48d2-a353-22b1f983191f/artifacts/018486f5-301a-4b92-baa4-b5ab1b896a6d/download
Checking security report: postgres_exporter-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e458-49d8-a5f8-577b3afc2f26/artifacts/018486f5-3b21-4e33-93f8-ee30b9990746/download
Checking security report: redis-store-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e45d-4136-ac94-96f333a50975/artifacts/018486f5-4add-4273-876b-c002ff5b9af1/download
Checking security report: postgres-12-alpine-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e457-4884-8d24-50f53a4340c4/artifacts/018486f5-5ce4-48bd-b63a-4c9d4f4a66a7/download
Checking security report: github-proxy-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e44d-42f8-9eef-16fa8b7644b9/artifacts/018486f5-6183-4626-a652-a9d614c7fa97/download
Checking security report: jaeger-agent-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e452-45d8-86c4-0026853f106c/artifacts/018486f5-61b9-4e8a-8038-610e3cc40bcf/download
Checking security report: minio-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e454-4e44-841e-44133538e431/artifacts/018486f5-61b6-45fb-a1e8-2304e24aedcc/download
Checking security report: alpine-3.14-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e447-4c03-b832-5978635ffebd/artifacts/018486f5-6b08-4808-90c5-51bc537f4406/download
Checking security report: grafana-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e44f-4fda-a30a-00938cad9ace/artifacts/018486f5-71b4-4cd9-9d6b-ca2dc205fb4d/download
Checking security report: codeintel-db-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e44b-4e2a-b9b6-0b22156fef00/artifacts/018486f5-8c9f-4f4e-a07c-68593024e3e4/download
Checking security report: executor-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e467-43e8-922f-88cb0e0eecfc/artifacts/018486f5-a4f6-4d4e-a249-88899177be8b/download
Checking security report: codeinsights-db-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e449-4a9a-a42e-1b0bebb45b50/artifacts/018486f5-cac8-4233-906a-6d57bf55f20e/download
Checking security report: prometheus-gcp-security-report.html https://api.buildkite.com/v2/organizations/sourcegraph/pipelines/sourcegraph/builds/184191/jobs/018486f2-e45b-420b-a7e2-b46da77308d5/artifacts/018486f5-cd6d-4f00-9702-2f3af2756695/download
Found 46 CVEs in the build
All CVEs approved!

```

## Test plan

Manually tested against the current RC build

### CODE REVIEW

File: dev/sg/release.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func downloadUrl(uri string, w io.Writer) (err error) {
+	std.Out.WriteLine(output.Styledf(output.StylePending, "Downloading url: %s", uri))
+	client := http.Client{}
+	resp, err := client.Get(uri)
+	if err != nil {
+		return err
+	}
+	defer func() { err = resp.Body.Close() }()
```
Comment (camdencheek) [inline dev/sg/release.go:133]: this will always clobber an error from `io.Copy` below, right?

File: dev/sg/release.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		std.Out.WriteLine(output.Styledf(output.StyleSuccess, "All CVEs approved!"))
+	}
+
+	return nil
+}
+
+func downloadUrl(uri string, w io.Writer) (err error) {
+	std.Out.WriteLine(output.Styledf(output.StylePending, "Downloading url: %s", uri))
+	client := http.Client{}
+	resp, err := client.Get(uri)
```
Comment (camdencheek) [inline dev/sg/release.go:129]:
```suggestion
	resp, err := http.Get(uri)
```

File: dev/sg/release.go (hunk 1 above)
Comment (coury-clark) [reply dev/sg/release.go:133]: oops - you're right, I'll fix that

File: dev/sg/release.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+				cve := strings.TrimSpace(match[1])
+				foundCVE = append(foundCVE, cve)
+				if !strings.Contains(hbPage, cve) {
+					unapprovedCVE = append(unapprovedCVE, cve)
+				}
+			}
+		}
+	}
+
+	// verbose := verboseFlag.Get(cmd)
```
Comment (camdencheek) [inline dev/sg/release.go:107]: Is this meant to be commented?

File: dev/sg/release.go (hunk 2 above)
Comment (coury-clark) [reply dev/sg/release.go]: https://github.com/sourcegraph/sourcegraph/pull/44585/commits/225b814d3c117c54b9e75a79bc55c0173380e472

File: dev/sg/release.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	var foundCVE []string
+	var unapprovedCVE []string
+
+	for _, artifact := range artifacts {
+		name := *artifact.Filename
+		url := *artifact.DownloadURL
+
+		pattern, err := regexp.Compile(`<\w+>(CVE.*)<\/\w+>`)
```
Comment (camdencheek) [inline dev/sg/release.go:82]: IIRC, `.*` is greedy by default. Won't this match something like `<abc>CVE</abc><def></def>` with the capture group `CVE</abc><def>`?

File: dev/sg/release.go (hunk 1 above)
Comment (coury-clark) [reply dev/sg/release.go]: Seems like it's a standard pattern to ignore the error in Close, so I've just done that https://github.com/sourcegraph/sourcegraph/pull/44585/commits/225b814d3c117c54b9e75a79bc55c0173380e472

File: dev/sg/release.go (hunk 4 above)
Comment (coury-clark) [reply dev/sg/release.go:82]: good call out - you're right. It seems the actual format is specified so I've changed the pattern to `<\w+>(CVE-\d+-\d+)<\/\w+>`

**Review by BolajiOlajide — APPROVED** [review]
Love that we're slowly moving part of the release process to the `sg` tool.

Thanks @coury-clark!

**Review by jhchabran — APPROVED** [review]
Nice! Thank you for taking the time to bring this in!

**Comment by jhchabran** [comment]
cc @davejrt as you're keeping an eye on release stuff being ported in `sg` :)

**Comment by unknwon** [comment]
Nice!

**Pull request merged**